File: 122.best-time-to-buy-and-sell-stock-ii/122.best-time-to-buy-and-sell-stock-ii.java
/*
 * @lc app=leetcode id=122 lang=java
 *
 * [122] Best Time to Buy and Sell Stock II
 */

// @lc code=start
class Solution {
    public int maxProfit(int[] prices) {
        int len = prices.length;
        int profit0 = 0, profit1 = Integer.MIN_VALUE;
        for(int p : prices) {
            profit0 = Math.max(profit0, profit1+p);
            profit1 = Math.max(profit1, profit0-p);
        }
        return profit0;
    }
}
// @lc code=end

File: algorithm/122.best-time-to-buy-and-sell-stock-ii/122.best-time-to-buy-and-sell-stock-ii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        # dp[i][0] not holding the stack
        # dp[i][1] holding the stock
        dp = [[0] * 2 for _ in range(n + 1)]
        dp[0][1] = float('-inf') # Integer min, max
        for i in range(1, n+1):
            price = prices[i-1]
            dp[i][0] = max(dp[i-1][1] + price, dp[i-1][0])
            dp[i][1] = max(dp[i-1][0] - price, dp[i-1][1])
        
        return dp[n][0]
